src/libopensc/card-jpki.c

```diff
@@ -0,0 +1,344 @@
+{
+	int rc;
+	sc_path_t path;
+
+	/* Select JPKI application */
+	sc_format_path("D392f000260100000001", &path);
+	path.type = SC_PATH_TYPE_DF_NAME;
+	rc = sc_select_file(card, &path, NULL);
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rc, "select JPKI AP failed");
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_match_card(struct sc_card *card)
+{
+	int i, rc;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	i = _sc_match_atr(card, jpki_atrs, &card->type);
+	if (i >= 0) {
+		return 1;
+	}
+	rc = jpki_select_ap(card);
+	if (rc == SC_SUCCESS) {
+		card->type = SC_CARD_TYPE_JPKI_BASE;
+		return 1;
+	}
+	return 0;
+}
+
+static int
+jpki_finish(sc_card_t * card)
+{
+	struct jpki_private_data *drvdata = DRVDATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (drvdata) {
+		free(drvdata);
+		card->drv_data = NULL;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_init(struct sc_card *card)
+{
+	struct jpki_private_data *drvdata;
+	sc_file_t *mf;
+	int flags;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	drvdata = malloc(sizeof (struct jpki_private_data));
+	if (!drvdata) {
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	memset(drvdata, 0, sizeof (struct jpki_private_data));
+	sc_format_path("D3 92 f0 00 26 01 00 00 00 01", &drvdata->aid);
+
+	/* create virtual MF */
+	mf = sc_file_new();
+	if (!mf) {
+		free(drvdata);
+		return SC_ERROR_OUT_OF_MEMORY;
+	}
+	sc_format_path("3f00", &mf->path);
+	mf->type = SC_FILE_TYPE_DF;
+	mf->shareable = 0;
+	mf->ef_structure = SC_FILE_EF_UNKNOWN;
+	mf->size = 0;
+	mf->id = 0x3f00;
+	mf->status = SC_FILE_STATUS_ACTIVATED;
+	sc_file_add_acl_entry(mf, SC_AC_OP_SELECT, SC_AC_NONE, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_LIST_FILES, SC_AC_NONE, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_LOCK, SC_AC_NEVER, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_DELETE, SC_AC_NEVER, 0);
+	sc_file_add_acl_entry(mf, SC_AC_OP_CREATE, SC_AC_NEVER, 0);
+	drvdata->mf = mf;
+
+	card->name = "jpki";
+	card->drv_data = drvdata;
+
+	flags = SC_ALGORITHM_RSA_HASH_NONE | SC_ALGORITHM_RSA_PAD_PKCS1;
+	_sc_card_add_rsa_alg(card, 2048, flags, 0);
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int
+jpki_select_file(struct sc_card *card,
+				 const struct sc_path *path, struct sc_file **file_out)
+{
+	struct jpki_private_data *drvdata = DRVDATA(card);
+	int rc;
+	sc_apdu_t apdu;
+	struct sc_file *file = NULL;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			 "jpki_select_file: path=%s, len=%d\n",
+			 sc_print_path(path), path->len);
+	if (path->len == 2 && memcmp(path->value, "\x3F\x00", 2) == 0) {
+		drvdata->selected = SELECT_MF;
+		if (file_out) {
+			sc_file_dup(file_out, drvdata->mf);
+			if (*file_out == NULL) {
+				return SC_ERROR_OUT_OF_MEMORY;
+			}
+		}
+		return 0;
+	}
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 0, 0);
+	switch (path->type) {
+	case SC_PATH_TYPE_FILE_ID:
+		apdu.p1 = 2;
```

Comment:
If this works, you could directly change the PATH type of your certificate objects in your PKCS#15 emulation...